src/compiler/parser.ts

```diff
@@ -1891,6 +1891,10 @@ namespace Parser {
                     }
 
                     if (pos >= 0) {
+                        // skip all statements in range already completely consumed by the reparsed statement
+                        while (sourceFile.statements[pos].end < statement.end) {
+                            pos++;
+                        }
                         const nonAwaitStatement = sourceFile.statements[pos];
                         if (statement.end === nonAwaitStatement.pos) {

```

Comment:
I would like to exploit the parallelism between the `while`, `if` and `else if` to make the mathematical properties more obvious. Specifically, drop `nonAwaitStatement` and flip the order of comparisons in the two `if`s so that `sourceFile.statements[pos]` is always first. That way you can see the `while` skipping over sourceFiles.statements, then the two ifs checking whether the pos is exactly equal to the end, or has gone beyond it.

(It, uh, also points out what a bad name `pos` and `start` are, since they're so similar to the different meetings of pos/end in TextRange.)

```suggestion
                        if (sourceFile.statements[pos].pos === statement.end) {
```